plugins/destination/plugin.go

```diff
@@ -3,54 +3,127 @@ package destination
 
+type ManagedWriter interface {
+	PreWriteTableBatch(ctx context.Context, tables schema.Tables) error
+	WriteTableBatch(ctx context.Context, table *schema.Table, data [][]interface{}) error
+}
+
+type UnimplementedManagedWriter struct{}
+
+type UnmanagedWriter interface {
+	Write(ctx context.Context, tables schema.Tables, res <-chan *ClientResource) error
+	Metrics() Metrics
+}
+
+type UnimplementedUnmanagedWriter struct{}
+
+func (*UnimplementedManagedWriter) PreWriteTableBatch(context.Context, schema.Tables) error {
+	panic("PreWriteTableBatch not implemented")
+}
+
+func (*UnimplementedManagedWriter) WriteTableBatch(context.Context, *schema.Table, [][]interface{}) error {
+	panic("WriteTableBatch not implemented")
+}
+
+func (*UnimplementedUnmanagedWriter) Write(context.Context, schema.Tables, <-chan *ClientResource) error {
+	panic("Write not implemented")
+}
+
+func (*UnimplementedUnmanagedWriter) Metrics() Metrics {
+	panic("Metrics not implemented")
+}
+
 type Client interface {
 	schema.CQTypeTransformer
-	ReverseTransformValues(table *schema.Table, values []any) (schema.CQTypes, error)
+	ReverseTransformValues(table *schema.Table, values []interface{}) (schema.CQTypes, error)
 	Migrate(ctx context.Context, tables schema.Tables) error
-	Read(ctx context.Context, table *schema.Table, sourceName string, res chan<- []any) error
-	Write(ctx context.Context, tables schema.Tables, res <-chan *ClientResource) error
-	Metrics() Metrics
+	Read(ctx context.Context, table *schema.Table, sourceName string, res chan<- []interface{}) error
+	ManagedWriter
+	UnmanagedWriter
 	DeleteStale(ctx context.Context, tables schema.Tables, sourceName string, syncTime time.Time) error
 	Close(ctx context.Context) error
 }
 
 type ClientResource struct {
 	TableName string
-	Data      []any
+	Data      []interface{}
 }
 
+type Option func(*Plugin)
+
 type Plugin struct {
 	// Name of destination plugin i.e postgresql,snowflake
 	name string
 	// Version of the destination plugin
 	version string
 	// Called upon configure call to validate and init configuration
-	newDestinationClient NewClientFunc
+	newClient NewClientFunc
+	writer    writerType
 	// initialized destination client
 	client Client
 	// spec the client was initialized with
 	spec specs.Destination
 	// Logger to call, this logger is passed to the serve.Serve Client, if not define Serve will create one instead.
 	logger zerolog.Logger
+
+	// This is in use if the user passed a managed client
+	metrics     map[string]*Metrics
+	metricsLock *sync.RWMutex
+
+	workers     map[string]*worker
+	workersLock *sync.Mutex
+
+	maxWorkers int
 }
 
-const writeWorkers = 1
+func WithManagerWriter() Option {
+	return func(p *Plugin) {
+		p.writer = managed
+	}
+}
 
-func NewPlugin(name string, version string, newDestinationClient NewClientFunc) *Plugin {
+func WithMaxWorkers(maxWorkers int) Option {
+	return func(p *Plugin) {
+		p.maxWorkers = maxWorkers
+	}
+}
+
+// NewPlugin creates a new unmanaged plugin
```

Comment:
I think this comment isn't strictly true (or at least complete), because the `WithManagerWriter` option can be passed in to make it use a managed writer?